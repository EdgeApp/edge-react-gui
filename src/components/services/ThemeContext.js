// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
