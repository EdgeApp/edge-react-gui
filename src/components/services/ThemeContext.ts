import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
