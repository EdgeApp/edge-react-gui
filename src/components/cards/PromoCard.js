// @flow

import * as React from 'react'
import { TouchableOpacity } from 'react-native'
import FastImage from 'react-native-fast-image'
import { TouchableWithoutFeedback } from 'react-native-gesture-handler'
import Animated, { interpolate, useAnimatedStyle, useSharedValue, withTiming } from 'react-native-reanimated'
import AntDesignIcon from 'react-native-vector-icons/AntDesign'

import { useHandler } from '../../hooks/useHandler.js'
import { type Theme, cacheStyles, getTheme } from '../services/ThemeContext.js'
import { EdgeText } from '../themed/EdgeText.js'

type Props = {
  message: string,
  iconUri?: string,
  animationValue: Animated.SharedValue<number>,
  onPress: () => void,
  onClose: () => void
}

export const PromoCard = (props: Props) => {
  const { message, iconUri, onPress, onClose, animationValue } = props
  const theme = getTheme()
  const styles = getStyles(theme)

  const pressed = useSharedValue(0)

  const handlePressIn = useHandler(() => (pressed.value = 1))
  const handlePressOut = useHandler(() => (pressed.value = 0))

  const borderAnimatedStyle = useAnimatedStyle(() => {
    return {
      borderColor: withTiming(pressed.value ? theme.promoCardBorderColorPressIn : theme.promoCardBorderColorPressOut, { duration: 100 }),
      opacity: interpolate(animationValue.value, [-1, 0, 1], [0.5, 1, 0.5])
    }
  })

  const contentAnimatedStyle = useAnimatedStyle(() => {
    return {
      opacity: withTiming(interpolate(animationValue.value, [-1, 0, 1], [0, 1, 0]), { duration: 10 })
    }
  })

  return (
    <TouchableWithoutFeedback onPressIn={handlePressIn} onPressOut={handlePressOut} activeOpacity={1} onPress={onPress}>
      <Animated.View style={[styles.container, borderAnimatedStyle]}>
        <Animated.View style={[styles.content, contentAnimatedStyle]}>
          {iconUri != null ? <FastImage resizeMode="contain" source={{ uri: iconUri }} style={styles.icon} /> : null}

          <EdgeText numberOfLines={0} style={styles.text}>
            {message}
          </EdgeText>

          <TouchableOpacity onPress={onClose}>
            <AntDesignIcon name="close" color={theme.iconTappable} size={theme.rem(2.5)} style={styles.close} />
          </TouchableOpacity>
        </Animated.View>
      </Animated.View>
    </TouchableWithoutFeedback>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    backgroundColor: theme.promoCardBackground,
    borderWidth: theme.promoCardBorderWidth,
    borderRadius: theme.rem(0.75),
    padding: theme.rem(1),
    height: '100%'
  },
  content: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center'
  },
  icon: {
    width: theme.rem(4),
    height: theme.rem(4)
  },
  text: {
    flex: 1,
    marginLeft: theme.rem(1),
    fontSize: theme.rem(3.5)
  },
  close: {}
}))
