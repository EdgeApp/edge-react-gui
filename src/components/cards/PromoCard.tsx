import * as React from 'react'
import { TouchableOpacity, View } from 'react-native'
import FastImage from 'react-native-fast-image'
import AntDesignIcon from 'react-native-vector-icons/AntDesign'

import { hideMessageTweak } from '../../actions/AccountReferralActions'
import { linkReferralWithCurrencies } from '../../actions/WalletListActions'
import { useHandler } from '../../hooks/useHandler'
import { lstrings } from '../../locales/strings'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { NavigationBase } from '../../types/routerTypes'
import { bestOfMessages } from '../../util/ReferralHelpers'
import { showError } from '../services/AirshipInstance'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { EdgeText } from '../themed/EdgeText'
import { ButtonBox } from '../themed/ThemedButtons'

interface Props {
  navigation: NavigationBase
}

export function PromoCard(props: Props) {
  const { navigation } = props
  const theme = useTheme()
  const styles = getStyles(theme)
  const dispatch = useDispatch()

  const accountMessages = useSelector(state => state.account.referralCache.accountMessages)
  const accountReferral = useSelector(state => state.account.accountReferral)
  const messageSummary = bestOfMessages(accountMessages, accountReferral)

  const handlePress = useHandler(() => {
    const uri = messageSummary?.message.uri
    if (uri != null) dispatch(linkReferralWithCurrencies(navigation, uri)).catch(err => showError(err))
  })

  const handleClose = useHandler(() => {
    if (messageSummary != null) {
      dispatch(hideMessageTweak(messageSummary.messageId, messageSummary.messageSource)).catch(err => showError(err))
    }
  })

  if (messageSummary == null) return null
  const { message } = messageSummary
  return (
    <ButtonBox marginRem={1} onPress={handlePress}>
      <View style={styles.container}>
        {message.iconUri != null ? <FastImage resizeMode="contain" source={{ uri: message.iconUri }} style={styles.icon} /> : null}
        <EdgeText testID="promoCard" numberOfLines={0} style={styles.text}>
          {message.message}
        </EdgeText>
        <TouchableOpacity accessible={false} onPress={handleClose}>
          <AntDesignIcon
            testID="closePromo"
            name="close"
            color={theme.iconTappable}
            size={theme.rem(1)}
            style={styles.close}
            accessibilityHint={lstrings.close_hint}
          />
        </TouchableOpacity>
      </View>
    </ButtonBox>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: theme.tileBackground,
    padding: theme.rem(0.5)
  },
  icon: {
    width: theme.rem(2),
    height: theme.rem(2),
    margin: theme.rem(0.5)
  },
  text: {
    flex: 1,
    margin: theme.rem(0.5)
  },
  close: {
    padding: theme.rem(0.5)
  }
}))
