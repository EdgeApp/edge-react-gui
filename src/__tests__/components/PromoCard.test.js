/* globals describe it expect */
/* eslint-disable flowtype/require-valid-file-annotation */

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { PromoCardComponent } from '../../components/cards/PromoCard.js'
import { getTheme } from '../../components/services/ThemeContext.js'

describe('PromoCard', () => {
  it('should render with loading props', () => {
    const renderer = new ShallowRenderer()

    const props = {
      accountMessages: ['MessageTweak'],
      accountReferral: {
        installerId: 'string',
        currencyCodes: ['BTC'],
        promotions: [],
        ignoreAccountSwap: true,
        hiddenAccountMessages: ['messageId']
      },

      hideMessageTweak: (messageId, source) => undefined,
      linkReferralWithCurrencies: uri => undefined,
      theme: getTheme()
    }
    const actual = renderer.render(<PromoCardComponent {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
