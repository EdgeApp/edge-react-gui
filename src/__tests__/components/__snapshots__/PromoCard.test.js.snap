// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PromoCard should render with loading props 1`] = `
<ButtonBox
  marginRem={1}
  onPress={[Function]}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(255, 255, 255, 0)",
        "borderColor": "rgba(255, 255, 255, .1)",
        "borderRadius": 4,
        "borderWidth": 1,
        "flexDirection": "row",
        "padding": 11,
      }
    }
  >
    <WithTheme(EdgeTextComponent)
      numberOfLines={0}
      style={
        Object {
          "flex": 1,
          "margin": 11,
        }
      }
    >
      MessageTweak
    </WithTheme(EdgeTextComponent)>
    <TouchableOpacity
      onPress={[Function]}
    >
      <Icon
        allowFontScaling={false}
        color="#66EDA8"
        name="close"
        size={22}
        style={
          Object {
            "padding": 11,
          }
        }
      />
    </TouchableOpacity>
  </View>
</ButtonBox>
`;
